text/0000-additivescenes-and-networkscenemanager-refactoring.md

```diff
@@ -0,0 +1,247 @@
+![](0000-additivescenes-and-networkscenemanager-refactoring/NSM_Reference2.png)
+
+As an example, there are two paths to accomplish the same general scene loading and synchronization process:
+* [OnFirstSceneSwitchSync ](https://github.com/Unity-Technologies/com.unity.multiplayer.mlapi/blob/6d9d31724f93db0fdac5ad6c76f65fc549746fe6/com.unity.multiplayer.mlapi/Runtime/SceneManagement/NetworkSceneManager.cs#L242)and [OnSceneSwitch ](https://github.com/Unity-Technologies/com.unity.multiplayer.mlapi/blob/6d9d31724f93db0fdac5ad6c76f65fc549746fe6/com.unity.multiplayer.mlapi/Runtime/SceneManagement/NetworkSceneManager.cs#L215)effectively accomplish the same thing, are handled in different locations, and contain the exact same code with two exceptions:
+    * [OnFirstSceneSwitchSync ](https://github.com/Unity-Technologies/com.unity.multiplayer.mlapi/blob/6d9d31724f93db0fdac5ad6c76f65fc549746fe6/com.unity.multiplayer.mlapi/Runtime/SceneManagement/NetworkSceneManager.cs#L242)checks to see if the client is already in the same scene and exists early.  If not it loads the scene synchronously (i.e. blocking until complete) and then notifies the client loaded the scene.
+    * [OnSceneSwitch ](https://github.com/Unity-Technologies/com.unity.multiplayer.mlapi/blob/6d9d31724f93db0fdac5ad6c76f65fc549746fe6/com.unity.multiplayer.mlapi/Runtime/SceneManagement/NetworkSceneManager.cs#L215)does not check to see if the client is in the same scene, it preserves NetworkObjects that don't have DestroyWithScene set to true, and then it loads the scene asynchronously.
+
+While this architecture worked with single mode scene loading, additive scene loading creates a new level of complexity that could become problematic to maintain a unified scene loading and synchronization process.  Consider the following diagram below:
+
+![](0000-additivescenes-and-networkscenemanager-refactoring/NSM_Reference8.png)
+
+With the existing netcode architecture NetworkObjects are always associated with the currently active scene and are serialized in no specific order nor grouped by any form of scene dependencies.  The above diagram outlines one problematic scenario where  NetworkPrefabHandler spawn generators are additively loaded in **Scene_A-1** and **Scene_B-2**.  On the server or host side, there would be no real issues as the scenes would be loaded in the appropriate order and the NetworkObjects would not need to be synchronized locally.  However, on the client side issues arise when NetworkObjects that were dependent upon either **Scene_A-1** or **Scene_B-2** are instantiated before their dependent additive scenes are loaded.   
```

Comment:
> However, on the client side issues arise when NetworkObjects that were dependent upon either **Scene_A-1** or **Scene_B-2** are instantiated before their dependent additive scenes are loaded.

I'm not sure whether this problem needs to be solved by us or not. I'd expect the developer who implements "spawn pools inside additively loaded scenes" to solve this problem for themselves. Though I might be narrowly-minded here.

Reply:
Yeah, I looked at various possible approaches for this by playing the role of a user and trying various scenarios where a user has a dependency for a NetworkObject that is instantiated in the current active scene (i.e. single loaded mode) by some user code (component) in a NetworkObject/GameObject within an additive scene.  I used NetworkObject pools that use custom Network Prefab handlers since that too was a new feature in this milestone (i.e. make sure the two feature play nice with each other).  Without the NetworkSceneManager knowing about any form of scene dependencies the issue only arises when a player is joining a game session in progress.  
The initial implementation that did not have any form of scene dependency functionality I tried several approaches from the "user code side" and it resulted in a very complex set of code the user would be required  to write to detect any NetworkObjects instantiated, determine if they belong to a NetworkObject pool, inserting the NetworkObject into the pool (and releasing the NetworkObject in the pool's position it took), but in the end it ended up being a high user pain threshold that didn't work "smoothly".

**The primary issue boils down to order of operations:** 
_When do you instantiate NetworkObjects and should other additive scenes be loaded first before those NetworkObjects are instantiated?_

By adding the ability to set a NetworkObject's scene dependency and making the NetworkSceneManager aware of this potential setting,  the user only has to make sure NetworkObjects that depend upon a specific scene to be loaded so the assets that a NetworkObject are dependent upon are instantiated before or with the dependent NetworkObjects.  

We don't completely automate this process for the user, but we provide the option to at least assure a NetworkObject is not instantiated until a specific scene is loaded.  

NetworkSceneManager only groups NetworkObjects to the scenes they are associated with or have a dependency with and will deserialize them in that order.  We do not migrate NetworkObjects to scenes or the like, that is left up to the user, but we do provide a way to assure that when a NetworkObject is being deserialized it is done only after the scene it is associated with (GameObject.scene) or it is dependent upon (NetworkObject.DependentSceneName) is loaded first.  What happens from there is up to the user.

Reply:
This was resolved by updating the way I approached how things are loaded.
Now we load all scenes first (not spawning or soft synchronizing during the loading), and then once all scenes are loaded we handle local spawning and soft synchronization.  Made updates to the RFC to reflect these changes.